src/libappimage/utils/IconHandle.cpp

```diff
@@ -0,0 +1,417 @@
+                // Icons are squared so we only have to query for one size value
+                if (imageFormat == "png" && cairoSurface != nullptr)
+                    return cairo.image_surface_get_height(cairoSurface);
+
+
+                if (imageFormat == "svg" && rsvgHandle != nullptr) {
+                    Priv::RSvgHandle::RsvgDimensionData dimensions = {};
+                    rsvg.handle_get_dimensions(rsvgHandle, &dimensions);
+
+                    return dimensions.height;
+                }
+
+                throw IconHandleError("Malformed IconHandle");
+            }
+
+            int getIconSize() const { return iconSize; }
+
+            void setIconSize(int iconSize) { Priv::iconSize = iconSize; }
+
+            const std::string& getImageFormat() const { return imageFormat; }
+
+            void save(const boost::filesystem::path& path, const std::string& targetFormat) {
+                const auto& output = getNewIconData(targetFormat);
+
+                if (output.empty())
+                    throw IconHandleError("Unable to transform " + imageFormat + " into " + targetFormat);
+
+                bf::ofstream ofstream(path, std::ios::out | std::ios::binary | std::ios::trunc);
+                if (ofstream.is_open())
+                    ofstream.write(reinterpret_cast<const char*>(output.data()), output.size());
+                else
+                    throw IconHandleError("Unable to write into: " + path.string());
+            }
+
+            std::vector<uint8_t> getNewIconData(const std::string& targetFormat) {
+                if (targetFormat == "png") {
+                    if (imageFormat == "svg")
+                        return svg2png();
+
+                    if (imageFormat == "png")
+                        return png2png();
+                }
+
+                if (targetFormat == "svg") {
+                    if (imageFormat == "svg")
+                        return originalData; // svgs doens't require to be resized
+
+                    if (imageFormat == "png")
+                        throw IconHandleError("png to svg conversion is not supported");
+                }
+            }
+        };
+
+        IconHandle::IconHandle(std::vector<uint8_t>& data) : priv(new Priv(data)) {}
+
+        int IconHandle::getSize() { return priv->getIconSize(); }
+
+        std::string IconHandle::format() { return priv->getImageFormat(); }
+
+        void IconHandle::setSize(int size) { priv->setIconSize(size); }
+
+        void IconHandle::save(const std::string& path, const std::string& format) {
+            bf::path bPath(path);
+            try { bf::create_directories(bPath.parent_path()); }
+            catch (const bf::filesystem_error &) { throw IconHandleError("Unable to create parent path"); }
+
+            priv->save(bPath, format);
+        }
+
+        IconHandle::IconHandle(const std::string& path) : priv(new Priv(path)) {}
+
+        IconHandle::~IconHandle() = default;
```

Comment:
In this particular scenario `dlclose()` is not required as `RTLD_NODELETE` is used on `dlopen` which means that the lib will not be unloaded on `dlclose`. It will be pointless to do it as we will use the same code several times during the execution of the application and also libcairo wasn't designed to be used this way therefore if we try to load it more than once the static data that it allocates gets corrupted and the whole app crashes.

Reply:
Why then declare a deconstructor (it's not at all needed here)? Also, this should be documented somewhere, I haven't seen this `RTLD_NODELETE` before.

Reply:
The destructor (`dlclose`) was added looking for completeness. At the time of writing that class, I didn't know that wasn't required to call `dlclose` in this particular scenario. 

Comments were added describing the functions of each mode flag and why they are used.